src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  FloatArray cs_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* cs = static_cast<FloatType*>(cs_array->data);
+  FloatArray ds_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* ds = static_cast<FloatType*>(ds_array->data);
+
+  IdxType max_d = 1;
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_d = indptr[rid + 1] - indptr[rid];
+    max_d = std::max(act_d, max_d);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_d;
+    cs[i] = num_picks / d;
+    ds[i] = d;
+    hop_size += act_d;
+  }
+
+  if (importance_sampling) {
+    FloatArray ps_array = NDArray::Empty({max_d + 1},
+                                          DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                          ctx);
+    FloatType* ps = static_cast<FloatType*>(ps_array->data);
+
+    double prev_ex_nodes = max_d * num_rows;
+
+    phmap::flat_hash_map<IdxType, FloatType> hop_map2;
+    for (int iters = 0; iters < importance_sampling || importance_sampling < 0; iters++) {
+      if (!weights || iters) {
+        hop_map2.clear();
+        for (int64_t i = 0; i < num_rows; ++i) {
+          const FloatType c = cs[i];
+          const IdxType rid = rows_data[i];
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+            const auto ct = c * (weights && iters == 1 ? A[j] : 1);
+            auto itb = hop_map2.emplace(indices[j], ct);
+            if (!itb.second)
+              itb.first->second = std::max(ct, itb.first->second);
+          }
+        }
+        if (hop_map.empty())
+          hop_map = std::move(hop_map2);
+        else
+          for (auto it : hop_map2)
+            hop_map[it.first] *= it.second;
+      }
+
+      for (int64_t i = 0; i < num_rows; ++i) {
+        const IdxType rid = rows_data[i];
+        const auto d = indptr[rid + 1] - indptr[rid];
+        if (d == 0)
+          continue;
+
+        const auto k = std::min(num_picks, d);
+
+        if (hop_map.empty()) {  // weights first iter, pi = A
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+            ps[j - indptr[rid]] = A[j];
+        } else {
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+            ps[j - indptr[rid]] = hop_map[indices[j]];
+        }
+
+        double var_target = ds[i] * ds[i] / k;
+        if (weights) {
+          var_target -= ds[i] * ds[i] / d;
+          for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+            var_target += A[j] * A[j];
+        }
+        FloatType c = cs[i];
+        double var_1;
+        if (linear_c_convergence) {
```

Comment:
If `linear_c_convergence` is `constexpr false`, why do we need the if-else block?